torchtext/datasets/language_modeling.py

```diff
@@ -1,217 +1,199 @@
-                subdirectory the data files will be stored.
-            wv_dir, wv_type, wv_dim: Passed to the Vocab constructor for the
-                text field. The word vectors are accessible as
-                train.dataset.fields['text'].vocab.vectors.
-            Remaining keyword arguments: Passed to the splits method.
         """
-        TEXT = data.Field()
-
-        train, val, test = cls.splits(TEXT, root=root, **kwargs)
-
-        TEXT.build_vocab(train, vectors=vectors)
-
-        return data.BPTTIterator.splits(
-            (train, val, test), batch_size=batch_size, bptt_len=bptt_len,
-            device=device)
 
+        super(LanguageModelingDataset, self).__init__()
+        self._data = data
+        self._vocab = vocab
+
+    def __getitem__(self, i):
+        return self._data[i]
+
+    def __len__(self):
+        return len(self._data)
+
+    def __iter__(self):
+        for x in self._data:
+            yield x
+
+    def get_vocab(self):
+        return self._vocab
+
+
+def _setup_datasets(dataset_name, tokenizer=get_tokenizer("basic_english"),
+                    root='.data', vocab=None, removed_tokens=['<unk>']):
+    if dataset_name == 'PennTreebank':
+        train_path = download_from_url(URLS['PennTreebank'][0], root=root)
+        test_path = download_from_url(URLS['PennTreebank'][1], root=root)
+        valid_path = download_from_url(URLS['PennTreebank'][2], root=root)
+    else:
+        dataset_tar = download_from_url(URLS[dataset_name], root=root)
+        extracted_files = extract_archive(dataset_tar)
+        for fname in extracted_files:
+            if 'train' in fname:
+                train_path = os.path.join(root, fname)
+            if 'test' in fname:
+                test_path = os.path.join(root, fname)
+            if 'valid' in fname:
+                valid_path = os.path.join(root, fname)
+
+    if vocab is None:
+        logging.info('Building Vocab based on {}'.format(train_path))
+        vocab = build_vocab_from_iterator(read_text_iterator(train_path, tokenizer))
+    else:
+        if not isinstance(vocab, Vocab):
+            raise TypeError("Passed vocabulary is not of type Vocab")
+    logging.info('Vocab has {} entries'.format(len(vocab)))
+    logging.info('Creating training data')
+    train_iter = create_data_from_iterator(
+        vocab, read_text_iterator(train_path, tokenizer), removed_tokens)
+    train_data = []
+    for tokens in train_iter:
+        train_data += tokens
+
+    logging.info('Creating testing data')
+    test_iter = create_data_from_iterator(
+        vocab, read_text_iterator(test_path, tokenizer), removed_tokens)
+    test_data = []
+    for tokens in test_iter:
+        test_data += tokens
+
+    logging.info('Creating valid data')
+    valid_iter = create_data_from_iterator(
+        vocab, read_text_iterator(valid_path, tokenizer), removed_tokens)
+    valid_data = []
+    for tokens in valid_iter:
+        valid_data += tokens
```

Comment:
nit: I don't know that it's valuable to simplify this block, but it'd be more readable by just adding a space between line 91 and 92 :)